cmd/schema.go

```diff
@@ -35,49 +34,75 @@ func (cmd *SchemaCmd) Synopsis() string {
-databases in direct connect mode can be specified by setting appropriate
-environment variables. The schema flags are:
+specified by either file param in source-profile or piped to stdin. Connection
+profile for source databases in direct connect mode can be specified by setting
+appropriate environment variables. The schema flags are:
 `, path.Base(os.Args[0]))
 }
 
 // SetFlags sets the flags.
 func (cmd *SchemaCmd) SetFlags(f *flag.FlagSet) {
+	f.StringVar(&cmd.source, "source", "", "Flag for specifying source DB, (e.g., `PostgreSQL`, `MySQL`, `DynamoDB`)")
+	f.StringVar(&cmd.sourceProfile, "source-profile", "", "Flag for specifying connection profile for source database e.g., \"file=<path>,format=dump\"")
+	f.StringVar(&cmd.target, "target", "Spanner", "Specifies the target DB, defaults to Spanner (accepted values: `Spanner`)")
+	f.StringVar(&cmd.targetProfile, "target-profile", "", "Flag for specifying connection profile for target database e.g., \"dialect=postgresql\"")
 	f.StringVar(&cmd.filePrefix, "prefix", "", "File prefix for generated files")
-	f.StringVar(&cmd.driverName, "driver", "pg_dump", "Flag for specifying source DB or dump files (accepted values are \"pg_dump\", \"postgres\", \"mysqldump\", \"mysql\", and \"dynamodb\")")
-	f.Int64Var(&cmd.schemaSampleSize, "schema-sample-size", int64(100000), "Number of rows to use for inferring schema (only for DynamoDB)")
-	f.StringVar(&cmd.dumpFilePath, "dump-file", "", "Location of dump file to process")
-	f.StringVar(&cmd.targetDb, "target-db", "", "Specifies the target DB, defaults to spanner")
 }
 
 func (cmd *SchemaCmd) Execute(ctx context.Context, f *flag.FlagSet, _ ...interface{}) subcommands.ExitStatus {
-	ioHelper := conversion.NewIOStreams(cmd.driverName, cmd.dumpFilePath)
+	sourceProfile, err := NewSourceProfile(cmd.sourceProfile, cmd.source)
+	if err != nil {
+		panic(err)
+	}
+	driverName, err := sourceProfile.ToLegacyDriver(cmd.source)
+	if err != nil {
+		panic(err)
+	}
+
+	targetProfile, err := NewTargetProfile(cmd.targetProfile)
+	if err != nil {
+		panic(err)
+	}
+	targetDb := targetProfile.ToLegacyTargetDb()
+
+	dumpFilePath := ""
+	if sourceProfile.ty == SourceProfileTypeFile && (sourceProfile.file.format == "" || sourceProfile.file.format == "dump") {
+		dumpFilePath = sourceProfile.file.path
+	}
+	ioHelper := conversion.NewIOStreams(driverName, dumpFilePath)
 	if ioHelper.SeekableIn != nil {
 		defer ioHelper.In.Close()
 	}
 
 	// If filePrefix not explicitly set, use generated dbName.
 	if cmd.filePrefix == "" {
-		dbName, err := conversion.GetDatabaseName(cmd.driverName, time.Now())
+		dbName, err := conversion.GetDatabaseName(driverName, time.Now())
 		if err != nil {
-			log.Fatalf("can't generate database name for prefix: %v\n", err)
+			panic(fmt.Errorf("can't generate database name for prefix: %v", err))
 		}
 		cmd.filePrefix = dbName + "."
 	}
 
+	schemaSampleSize := int64(100000)
```

Comment:
wondering if it makes sense to push the profile parsing logic into db specific functions or modules. There will be some duplication of code, but we can emit better error messages and separate out the code more cleanly

Reply:
That's the intention, but we would be able to do that in follow up refactoring when we pass around SourceProfile itself instead of passing around multiple arguments to conversion.SchemaConv and conversion.DataConv etc. Until then, this check needs to live here.